File: algorithms/0020_valid-parentheses/java/Solution1.java
import java.util.Stack;
class Solution1 {
    public boolean isValid(String s){
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c= s.charAt(i);
            // 压入栈
            if(c=='('||c=='['||c=='{'){
                stack.push(c);
            }else {
                // 若输入的是"]"
                if(stack.isEmpty()){
                    return false;
                }
                char top = stack.pop();
                if(top=='('&&c!=')'){
                    return false;
                }else if(top=='{'&&c!='}'){
                    return false;
                }else if(top=='['&&c!=']'){
                    return false;
                }
            }
        }
        return stack.isEmpty();
    }
    public static void main(String[] args) {
        System.out.println((new Solution1()).isValid(""));
        System.out.println((new Solution1()).isValid("([)]"));
    }
}

File: leetcode/algorithms/0020_valid-parentheses/cpp/Solution1.cpp
#include <iostream>
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

class Solution1 {
public:
  bool isValid(string s) {
    stack<char> first;
    for (int i = 0; i < s.size(); ++i) {
      cout << s[i] << endl;
      if (s[i] == '(' || s[i] == '{' || s[i] == '[') {
        first.push(s[i]);
      } else {
        // 若是"["
        if (first.empty()) {
          return false;
        }
        if (s[i] == ')' && first.top() != '(') {
          return false;
        }
        if (s[i] == ']' && first.top() != '[') {
          return false;
        }
        if (s[i] == '}' && first.top() != '{') {
          return false;
        }
        first.pop();
      }
    }
    return first.empty();
  }
};

int main() {
  cout << Solution1().isValid("[]");
  return 0;
}